### PR TITLE
networkd: Static neighbor support

### PR DESCRIPTION
**Pull request by wkennington**
When using networkd we currently have no way of ensuring that static
neighbor entries are set when our link comes up. This change adds a new
section to the network definition that allows multiple static neighbors
to be set on a link.

### CODE REVIEW

File: man/systemd.network.xml [hunk 1]
```diff
@@ -907,6 +907,28 @@
       </variablelist>
   </refsect1>
 
+  <refsect1>
+    <title>[Neighbor] Section Options</title>
+      <para>A <literal>[Neighbor]</literal> section accepts the
+      following keys. Specify several <literal>[Neighbor]</literal>
+      sections to configure several static neighbors.</para>
```
Comment (poettering) [inline man/systemd.network.xml]: this needs a longer explanation. Should mention the word "ARP" for IPv4, and maybe something about "this adds static entries to the neighbor table which establishes bindings between protocol addresses and link layer addresses for hosts sharing the same link" (based on the description in the ip-neighbor(8) man page)

Comment (poettering) [reply man/systemd.network.xml]: Also, mention that this is IPv4 and IPv6 specific

File: src/network/networkd-link.c [hunk 2]
```diff
@@ -910,6 +914,61 @@ int link_route_remove_handler(sd_netlink *rtnl, sd_netlink_message *m, void *use
...
+        assert(link->neighbor_messages > 0);
+
+        link->neighbor_messages--;
+
+        if (IN_SET(link->state, LINK_STATE_FAILED, LINK_STATE_LINGER))
+                return 1;
+
+        r = sd_netlink_message_get_errno(m);
+        if (r < 0 && r != -EEXIST)
+                log_link_warning_errno(link, r, "could not set neighbor: %m");
```
Comment (poettering) [inline src/network/networkd-link.c]: "could" → "Could"

File: src/network/networkd-link.h [hunk 3]
```diff
@@ -97,6 +99,8 @@ typedef struct Link {
         bool ipv4ll_address:1;
         bool ipv4ll_route:1;
 
+        bool neighbors_configured;
+
```
Comment (poettering) [inline src/network/networkd-link.h:102]: why the empty line, it makes sense to keep this together with the other variables `static_routes_configured` and such

File: src/network/networkd-neighbor.c [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        int r;
+
+        assert(neighbor);
+        assert(link);
+        assert(link->ifindex > 0);
+        assert(link->manager);
+        assert(link->manager->rtnl);
+
+        if (neighbor->family == AF_UNSPEC)
+                return log_error_errno(-EINVAL, "Missing Neighbor Address");
```
Comment (poettering) [inline src/network/networkd-neighbor.c]: `log_error_errno(SYNTHETIC_ERRNO(EINVAL),  …)` please, see https://systemd.io/CODING_STYLE

File: src/network/networkd-neighbor.c [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+        if (!neighbor->mac_configured)
+                return log_error_errno(-EINVAL, "Missing Neighbor MACAddress");
```
Comment (poettering) [inline src/network/networkd-neighbor.c]: as above

Comment (poettering) [reply src/network/networkd-neighbor.c]: also, these two error messages could have clearer wording... "Neighbor without MACAddress= configured" or so

File: src/network/networkd-neighbor.c [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+                if (r < 0)
+                        return log_error_errno(r, "Could not append NDA_DST attribute: %m");
+                break;
+        case AF_INET:
+                r = sd_netlink_message_append_in_addr(req, NDA_DST, &neighbor->in_addr.in);
+                if (r < 0)
+                        return log_error_errno(r, "Could not append NDA_DST attribute: %m");
+                break;
+        default:
+                return log_error_errno(-EINVAL, "Invalid neighbor address type");
```
Comment (poettering) [inline src/network/networkd-neighbor.c]: as above

File: src/network/networkd-neighbor.c [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+                return r;
+
+        r = in_addr_from_string_auto(rvalue, &f, &buffer);
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r, "Neighbor Address is invalid, ignoring assignment: %s", rvalue);
+                return 0;
+        }
+
+        n->family = f;
+        n->in_addr = buffer;
```
Comment (poettering) [inline src/network/networkd-neighbor.c]: why the indirection through f/buffer? in_addr_from_string_auto() could directly write to these fields of `n`, no?

File: src/network/networkd-neighbor.c [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+        if (r < 0)
+                return r;
+
+        r = ether_addr_from_string(rvalue, &buffer);
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r, "Neighbor MACAddress is invalid, ignoring assignment: %s", rvalue);
+                return 0;
+        }
+
+        n->mac = buffer;
```
Comment (poettering) [inline src/network/networkd-neighbor.c]: as above, `buffer` is not necessary, no?

File: test/test-network/systemd-networkd-tests.py [hunk 9]
```diff
@@ -658,6 +658,17 @@ def test_ipv6_address_label(self):
...
+    def test_ipv6_neighbor(self):
+        self.copy_unit_to_networkd_unit_path('25-neighbor-section.network', '12-dummy.netdev')
+        self.start_networkd()
+
+        self.assertTrue(self.link_exits('dummy98'))
+
+        output = subprocess.check_output(['ip', 'neigh', 'list']).rstrip().decode('utf-8')
+        print(output)
+        self.assertRegex(output, '192.168.10.1.*00:00:5e:00:02:65.*PERMANENT')
+        self.assertRegex(output, '2004:da8:1:0::1.*00:00:5e:00:02:66.*PERMANENT')
```
Comment (poettering) [inline test/test-network/systemd-networkd-tests.py:788]: thanks for adding a test, but any chance you can add one for ipv4 arp, too?

**Comment by poettering** [comment]
looks great, but a few (mostly superficial) comments, above

File: src/network/networkd-link.c [hunk 10]
```diff
@@ -935,7 +994,7 @@ static int address_handler(sd_netlink *rtnl, sd_netlink_message *m, void *userda
 
         if (link->address_messages == 0) {
                 log_link_debug(link, "Addresses set");
-                link_enter_set_routes(link);
+                link_enter_set_neighbors(link);
```
Comment (ssahani) [inline src/network/networkd-link.c]: Now we are setting address then routes. I think we should not touch this path. It should be done like ```address label```, ```ip routing rule``` independent of address then routes.  Any reason why  ```address->neighbor->routes``` ?

Comment (wak-google) [reply src/network/networkd-link.c]: The only reason I did it this way was to keep everything here simple. Theoretically static neighbor setting could happen in parallel with addresses and routes since there is no dependency either way on neighbors.

File: test/test-network/systemd-networkd-tests.py (hunk 9 above)
Comment (wkennington) [reply test/test-network/systemd-networkd-tests.py:788]: This does cover ipv4 arp with the 192.168.10.1 address. The neighbor table is a generic interface over ipv6 neighbor discovery and ipv4 arp.

You can verify this using `arp -n` and comparing to `ip -4 neigh`. `ip neigh` just shows both v4 and v6 together.

File: src/network/networkd-neighbor.c (hunk 7 above)
Comment (wkennington) [reply src/network/networkd-neighbor.c]: The main reason is that the underlying function inet_pton(3) does not guarantee that it won't overwrite the buffer on failure. If we specify two addresses in the config file and the second one is bad and inet_pton overwrites the buffer, we won't end up with a bad address in the output `struct Neighbor`. It does seem silly that inet_pton would clobber the `dst` when an error occurs, but I'm just conforming to the pattern used in other places. An example of similar behavior is in src/networkd/networkd-route.

File: src/network/networkd-neighbor.c (hunk 8 above)
Comment (wkennington) [reply src/network/networkd-neighbor.c]: as above

File: src/network/networkd-neighbor.c (hunk 4 above)
Comment (wkennington) [reply src/network/networkd-neighbor.c]: Done

File: src/network/networkd-neighbor.c (hunk 5 above)
Comment (wkennington) [reply src/network/networkd-neighbor.c]: I like that, fixed

File: src/network/networkd-neighbor.c (hunk 6 above)
Comment (wkennington) [reply src/network/networkd-neighbor.c]: Fixed

File: src/network/networkd-link.c (hunk 2 above)
Comment (wkennington) [reply src/network/networkd-link.c]: Fixed

File: src/network/networkd-link.h (hunk 3 above)
Comment (wkennington) [reply src/network/networkd-link.h:102]: Since configuring neighbors is its own stage it looks like it should be separate. The static_routes_configured is separate from dhcp{4,6}_configured and ndisc_configured is in its own section.

File: man/systemd.network.xml (hunk 1 above)
Comment (wak-google) [reply man/systemd.network.xml]: This should be clearer now.

File: src/network/networkd-link.c (hunk 10 above)
Comment (wkennington) [reply src/network/networkd-link.c]: I made some changes so we now have
`addresses -> routes` and `neightbors`
in parallel

File: man/systemd.network.xml [hunk 11]
```diff
@@ -907,6 +907,31 @@
...
+      entry to the neighbor table (IPv6) or arp table (IPv4) for
+      the given hardware address on the links matched for the network.
+      Specify several <literal>[Neighbor]</literal> sections to configure
+      several static neighbors.</para>
+
+      <variablelist class='network-directives'>
+        <varlistentry>
+          <term><varname>Address=</varname></term>
+          <listitem>
+            <para>The IP address of the neighbor</para>
```
Comment (yuwata) [inline man/systemd.network.xml]: Please add a period at the end.

File: man/systemd.network.xml [hunk 12]
```diff
@@ -907,6 +907,31 @@
...
+          </listitem>
+        </varlistentry>
+        <varlistentry>
+          <term><varname>MACAddress=</varname></term>
+          <listitem>
+            <para>The hardware address of the neighbor</para>
```
Comment (yuwata) [inline man/systemd.network.xml]: Same as the above.

File: src/network/networkd-neighbor.c [hunk 13]
```diff
@@ -0,0 +1,215 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include "alloc-util.h"
+#include "ether-addr-util.h"
+#include "conf-parser.h"
```
Comment (yuwata) [inline src/network/networkd-neighbor.c:6]: Please sort headers.

File: src/network/networkd-neighbor.c [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+#include "hashmap.h"
+#include "in-addr-util.h"
+#include "networkd-manager.h"
+#include "networkd-neighbor.h"
+#include "sd-netlink.h"
```
Comment (yuwata) [inline src/network/networkd-neighbor.c]: Please locate "sd-netlink.h" at the head of the headers, like
```
#include "sd-netlink.h"

#include "alloc-util.h"
...
```

File: src/network/networkd-neighbor.h [hunk 15]
```diff
@@ -0,0 +1,36 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+#pragma once
+
+#include "conf-parser.h"
+#include "ether-addr-util.h"
+#include "in-addr-util.h"
+#include "list.h"
+#include "macro.h"
+#include "sd-netlink.h"
```
Comment (yuwata) [inline src/network/networkd-neighbor.h]: Also here.

File: src/network/networkd-link.c [hunk 16]
```diff
@@ -910,6 +918,61 @@ int link_route_remove_handler(sd_netlink *rtnl, sd_netlink_message *m, void *use
         return 1;
 }
 
+static int neighbor_handler(sd_netlink *rtnl, sd_netlink_message *m, void *userdata) {
```
Comment (yuwata) [inline src/network/networkd-link.c]: I'd like to move `neighbor_handler()` to `networkd-neightbor.c`, and make `neighbor_configure()` take `NULL` as the callback. See #10976.

File: src/network/networkd-neighbor.c (hunk 7 above)
Comment (yuwata) [reply src/network/networkd-neighbor.c]: > The main reason is that the underlying function inet_pton(3) does not guarantee that it won't overwrite the buffer on failure.

https://github.com/kraj/glibc/blob/02cd5c1a8d033d7f91fea12a66bb44d1bbf85f76/resolv/inet_pton.c#L75-L78
https://github.com/kraj/glibc/blob/02cd5c1a8d033d7f91fea12a66bb44d1bbf85f76/resolv/inet_pton.c#L138-L140

File: src/network/networkd-neighbor.c (hunk 8 above)
Comment (yuwata) [reply src/network/networkd-neighbor.c]: `ether_addr_from_string()` does not update the buffer on failure.

File: src/network/networkd-link.c (hunk 16 above)
Comment (wkennington) [reply src/network/networkd-link.c]: I've updated this to follow the default callback convention but not rebasing on that commit unless it gets merged before this one. I'll update the async netlink call if that gets in.

**Comment by wkennington** [comment]
Updated, PTAL

File: src/network/networkd-neighbor.c (hunk 7 above)
Comment (yuwata) [reply src/network/networkd-neighbor.c]: I've created #11030.

File: src/network/networkd-link.c [hunk 17]
```diff
@@ -2385,7 +2384,7 @@ static int link_enter_join_netdev(Link *link) {
         assert(link->network);
         assert(link->state == LINK_STATE_PENDING);
 
-        link_set_state(link, LINK_STATE_ENSLAVING);
+        link_set_state(link, LINK_STATE_CONFIGURING);
```
Comment (poettering) [inline src/network/networkd-link.c:2349]: hmm? this should remain ENSLAVING, no?

I mean, you are merging SETTING_ADDRESSES and SETTING_ROUTES, but ENSLAVING should remain the way it is, no?

File: src/network/networkd-link.c [hunk 18]
```diff
@@ -3581,7 +3580,7 @@ int link_update(Link *link, sd_netlink_message *m) {
 
         if (link->state == LINK_STATE_LINGER) {
                 log_link_info(link, "Link readded");
-                link_set_state(link, LINK_STATE_ENSLAVING);
+                link_set_state(link, LINK_STATE_CONFIGURING);
```
Comment (poettering) [inline src/network/networkd-link.c:3545]: same here

File: src/network/networkd-link.c [hunk 19]
```diff
@@ -862,6 +865,8 @@ static int link_enter_set_routes(Link *link) {
```
Comment (poettering) [inline src/network/networkd-link.c:860]: The function link_enter_set_routes() is a bit of a misnomer as soon as there's no state SETTING_ROUTES anymore

File: src/network/networkd-link.c [hunk 20]
```diff
@@ -916,6 +920,34 @@ int link_route_remove_handler(sd_netlink *rtnl, sd_netlink_message *m, void *use
         return 1;
 }
 
+static int link_enter_set_neighbors(Link *link) {
```
Comment (poettering) [inline src/network/networkd-link.c]: the idea was always that functions called `…_enter_…` would enter a state. If there is no state associated with this, then they should have a different name. Maybe "request", "submit", "enqueue" or so would be a better word.

File: src/network/networkd-neighbor.c [hunk 21]
```diff
@@ -0,0 +1,243 @@
...
+
+        neighbor = new(Neighbor, 1);
+        if (!neighbor)
+                return -ENOMEM;
+
+        *neighbor = (Neighbor) {
+                .network = network,
+                .family = AF_UNSPEC,
+                .mac_configured = false,
+        };
```
Comment (poettering) [inline src/network/networkd-neighbor.c:62]: not that it matters, but the assignment of `.mac_configured` is redundant, as structured initialization initializes everything to 0 anyway that is not specified. (And yes, assigning of `.family` is strictly speaking also unnecessary, since `AF_UNSPEC` is `0`, but I think that one should be kept, as the fact that this is `0` is not as obvious as it is that `false` is `0`, if you follow what I mean)

File: src/network/networkd-neighbor.c (hunk 7 above)
Comment (poettering) [reply src/network/networkd-neighbor.c]: @wkennington `in_addr_from_string_auto()` really should be abstracting the weirdness of inet_pton() away. It is in a way a typesafe version of it, after all.

File: src/network/networkd-neighbor.c (hunk 8 above)
Comment (poettering) [reply src/network/networkd-neighbor.c]: (in fact, we have the general rule that our functions shall not modify output parameters on failure, and shall update all output parameters on success, see https://systemd.io/CODING_STYLE)

File: test/test-network/systemd-networkd-tests.py (hunk 9 above)
Comment (poettering) [reply test/test-network/systemd-networkd-tests.py:788]: oops, ignore me, i  was blind. didn't see the ipv4 addres there...

File: man/systemd.network.xml [hunk 22]
```diff
@@ -907,6 +907,31 @@
       </variablelist>
   </refsect1>
 
+  <refsect1>
+    <title>[Neighbor] Section Options</title>
+      <para>A <literal>[Neighbor]</literal> section accepts the
+      following keys. The neighbor section adds a permanent, static
+      entry to the neighbor table (IPv6) or arp table (IPv4) for
```
Comment (poettering) [inline man/systemd.network.xml]: acronyms are usally uppercased in our docs, i.e. `arp` → `ARP`, please

File: src/network/networkd-neighbor.c (hunk 7 above)
Comment (wkennington) [reply src/network/networkd-neighbor.c]: Done

File: src/network/networkd-neighbor.c (hunk 8 above)
Comment (wkennington) [reply src/network/networkd-neighbor.c]: Done

File: src/network/networkd-link.c (hunk 17 above)
Comment (wkennington) [reply src/network/networkd-link.c:2349]: Enslaving is treated as a stage of configuring the link. Any ordering between the sub-stages of configuring should be handled internally. Also, nothing conditionally depends on the value of enslaving and the mapping of this enum to strings called this "configuring" just like SETTING_*

File: src/network/networkd-neighbor.c (hunk 21 above)
Comment (wkennington) [reply src/network/networkd-neighbor.c:62]: Fair, I only added this for the clarity of someone searching the codebase.

File: man/systemd.network.xml (hunk 22 above)
Comment (wkennington) [reply man/systemd.network.xml]: Done

File: src/network/networkd-link.c (hunk 20 above)
Comment (wkennington) [reply src/network/networkd-link.c]: Any of these sound good, but I went with request.

File: src/network/networkd-link.c (hunk 19 above)
Comment (wkennington) [reply src/network/networkd-link.c:860]: Done

**Comment by wkennington** [comment]
Rebased on the latest changes by yuwata@ and addressed the above. PTAL

File: src/network/networkd-link.c [hunk 23]
```diff
@@ -909,6 +914,7 @@ static int address_handler(sd_netlink *rtnl, sd_netlink_message *m, Link *link)
 
         if (link->address_messages == 0) {
                 log_link_debug(link, "Addresses set");
+                link->addresses_configured = true;
                 link_enter_set_routes(link);
         }
```
Comment (poettering) [inline src/network/networkd-link.c:951]: it might be nice to eventually split out the five lines above into a new call `link_address_test_ready()` or so, which we can then call both from `address_handler()` and from `link_enter_set_addresses()`, since both execute kinda the same steps.

(But it's OK to leave this for now, and do that in another PR)

File: src/network/networkd-link.c (hunk 17 above)
Comment (poettering) [reply src/network/networkd-link.c:2349]: not sure i follow... why have LINK_STATE_ENSLAVING at all then?

Comment (wkennington) [reply src/network/networkd-link.c:2349]: What I was trying to say is that enslaving is just a part of link configuration that is no different from the address / route / neighbor configuration. It could actually be done in parallel with those operations since it requires the bond / bridge to be up. It's not a unique state like COMPLETED / FAILED / INVALID / UNMANAGED / PENDING.

**Comment by wkennington** [comment]
Hopefully I addressed what is left, PTAL

File: src/network/networkd-link.c [hunk 24]
```diff
@@ -743,6 +744,9 @@ void link_check_ready(Link *link) {
         if (!link->addresses_configured)
                 return;
 
+        if (!link->neighbors_configured)
+                return;
```
Comment (ssahani) [inline src/network/networkd-link.c:749]: we should consider this long time issue too . https://github.com/systemd/systemd/issues/9077

File: src/network/networkd-link.h [hunk 25]
```diff
@@ -20,8 +20,7 @@
 typedef enum LinkState {
         LINK_STATE_PENDING,
         LINK_STATE_ENSLAVING,
-        LINK_STATE_SETTING_ADDRESSES,
-        LINK_STATE_SETTING_ROUTES,
+        LINK_STATE_CONFIGURING,
```
Comment (poettering) [inline src/network/networkd-link.h:22]: you dropped "ENSLAVING" from link_state_table, but not from the enum. Needs to be dropped at both places

Comment (poettering) [inline src/network/networkd-link.h:22]: you dropped "ENSLAVING" from link_state_table, but not from the enum. Needs to be dropped at both places

Comment (wkennington) [reply src/network/networkd-link.h:22]: Good catch, done

**Comment by poettering** [comment]
lgtm. thanks!
